Automattic/simplenote-android
# Add App Log

### Fix
Add application logging to assist in troubleshooting data loss reports as described in https://github.com/Automattic/simplenote-android/issues/992 and https://github.com/Automattic/simplenote-android/issues/1101 among other issues.  The logs are not saved to device storage, but instead are stored in memory while the app is alive.  Once the app is killed by the operating system, the logs are cleared.  The logs are queued with a maximum of one hundred events to keep the memory footprint small.  The logs remain on the device and are never sent to anywhere unless explicitly done so by the user.  These aspects ensure an opt-in behavior and keeps the user in control of their information.  In order to share the logs, the **_Send logs_** item was added to the **_Settings_** screen.  When the **_Send logs_** item is tapped, a **_Share_** sheet is shown allowing the user to share the logs however they choose.  Since this feature is intended to help diagnose issues with data loss while editing/creating notes, the logs are focused on editor and syncing events.  See the details below for example logs.

<details>
<summary>
Log
</summary>
<pre>
Device: Google Pixel 4 (flame)
System: Android 10 (29)
App: Simplenote 2.9-rc-1 (111)<br>
Email: heckofatest@gmail.com
Notes: 21
Tags: 5<br>
2020-08-18 13:51:08.453 - LAYOUT: Small, portrait
2020-08-18 13:51:08.510 - NETWORK: WIFI (780.00 Mbps)
2020-08-18 13:51:08.511 - SCREEN: Created (NotesActivity)
2020-08-18 13:51:08.609 - NETWORK: WIFI (780.00 Mbps)
2020-08-18 13:51:08.610 - SCREEN: Created (NoteListFragment)
2020-08-18 13:51:08.662 - ACTION: App opened
2020-08-18 13:51:08.662 - SYNC: Started note bucket (NotesActivity)
2020-08-18 13:51:08.663 - SYNC: Started tag bucket (NotesActivity)
2020-08-18 13:51:10.155 - SCREEN: Paused (NoteListFragment)
2020-08-18 13:51:10.156 - SYNC: Stopped tag bucket (NotesActivity)
2020-08-18 13:51:10.157 - SYNC: Stopped note bucket (NotesActivity)
2020-08-18 13:51:10.157 - SCREEN: Paused (NotesActivity)
2020-08-18 13:51:10.171 - NETWORK: MOBILE (53.14 Mbps)
2020-08-18 13:51:10.171 - SCREEN: Created (NoteEditorActivity)
2020-08-18 13:51:10.218 - NETWORK: MOBILE (53.14 Mbps)
2020-08-18 13:51:10.218 - SCREEN: Resumed (NoteEditorActivity)
2020-08-18 13:51:10.231 - NETWORK: MOBILE (53.14 Mbps)
2020-08-18 13:51:10.231 - SCREEN: Created (NoteEditorFragment)
2020-08-18 13:51:10.262 - ACTION: Opened from list (NoteEditorFragment)
2020-08-18 13:51:10.265 - SCREEN: Created (NoteMarkdownFragment)
2020-08-18 13:51:10.266 - SYNC: Loaded note (Title: Logs / Characters: 6 / Words: 1)
2020-08-18 13:51:10.369 - SYNC: Started note bucket (NoteEditorFragment)
2020-08-18 13:51:14.243 - ACTION: Edited note (Title: Logs / Characters: 6 / Words: 1)
2020-08-18 13:51:14.260 - SYNC: Saved note (Title: Logs / Characters: 10 / Words: 2)
2020-08-18 13:51:16.712 - SCREEN: Paused (NoteEditorFragment)
2020-08-18 13:51:16.718 - SCREEN: Paused (NoteEditorActivity)
2020-08-18 13:51:16.740 - ACTION: Edited note (Title: Logs / Characters: 10 / Words: 2)
2020-08-18 13:51:17.176 - ACTION: App closed
2020-08-18 13:51:19.107 - ACTION: App opened
2020-08-18 13:51:19.109 - NETWORK: WIFI (650.00 Mbps)
2020-08-18 13:51:19.110 - SCREEN: Resumed (NoteEditorActivity)
2020-08-18 13:51:19.110 - SYNC: Started note bucket (NoteEditorFragment)
2020-08-18 13:51:20.595 - LAYOUT: Small, landscape
2020-08-18 13:51:22.667 - SYNC: Saved note (Title: Logs / Characters: 6 / Words: 1)
2020-08-18 13:51:22.668 - SCREEN: Paused (NoteEditorFragment)
2020-08-18 13:51:22.672 - SCREEN: Paused (NoteEditorActivity)
2020-08-18 13:51:22.679 - SYNC: Stopped note bucket (NoteEditorFragment)
2020-08-18 13:51:22.679 - ACTION: Edited note (Title: Logs / Characters: 6 / Words: 1)
2020-08-18 13:51:23.272 - ACTION: App closed
2020-08-18 13:51:24.050 - LAYOUT: Small, portrait
2020-08-18 13:51:27.204 - SYNC: Stopped note bucket (Simplenote)
2020-08-18 13:51:27.207 - SYNC: Stopped tag bucket (Simplenote)
2020-08-18 13:51:28.603 - ACTION: App opened
2020-08-18 13:51:28.605 - NETWORK: WIFI (650.00 Mbps)
2020-08-18 13:51:28.605 - SCREEN: Resumed (NoteEditorActivity)
2020-08-18 13:51:28.607 - SYNC: Started note bucket (NoteEditorFragment)
2020-08-18 13:51:29.307 - ACTION: Tapped back arrow in app bar (NoteEditorFragment)
2020-08-18 13:51:29.322 - SCREEN: Paused (NoteEditorFragment)
2020-08-18 13:51:29.323 - SCREEN: Paused (NoteEditorActivity)
2020-08-18 13:51:29.333 - ACTION: Edited note (Title: Logs / Characters: 6 / Words: 1)
2020-08-18 13:51:29.338 - SYNC: Started note bucket (NotesActivity)
2020-08-18 13:51:29.340 - SYNC: Started tag bucket (NotesActivity)
2020-08-18 13:51:29.844 - SYNC: Stopped note bucket (NoteEditorFragment)
2020-08-18 13:51:29.845 - SCREEN: Destroyed (NoteEditorFragment)
2020-08-18 13:51:29.848 - SYNC: Stopped note bucket (NoteMarkdownFragment)
2020-08-18 13:51:29.848 - SCREEN: Destroyed (NoteMarkdownFragment)
2020-08-18 13:51:31.653 - SCREEN: Paused (NoteListFragment)
2020-08-18 13:51:31.654 - SYNC: Stopped tag bucket (NotesActivity)
2020-08-18 13:51:31.654 - SYNC: Stopped note bucket (NotesActivity)
2020-08-18 13:51:31.654 - SCREEN: Paused (NotesActivity)
</pre>
</details>

There are seven types of log events based on the information of the event.  See the table below for details.

Type|Description
-|-
`ACCOUNT`|The user's email address, number of notes, and number of tags for the logged in account.
`ACTION`|The behavior performed by the user (e.g. editing a note, tapping back arrow).
`DEVICE`|The make, model, Android version, and Simplenote version installed on the device.
`LAYOUT`|The display size and orientation of the device.
`NETWORK`|The type and speed of the network connection.
`SCREEN`|The screen shown in the app (e.g. Note List, Note Editor).
`SYNC`|The events related to syncing (e.g. bucket start/stop, saving a note).

All the strings in these changes are intentionally hard-coded since none of those strings are user-facing and do not require translation.  Also, adding them to `strings.xml` and having them translated may cause confusion and difficult diagnosis when they are read by support or developers since they may not know the language.  The one exception is [`send_logs`](https://github.com/Automattic/simplenote-android/compare/add-app-log?expand=1#diff-19bb70917c478f507b793b466bd7d82eR34), which is user-facing and should be translated.

### Test
Since the application logging addresses multiple screens, the best testing method is to perform various tasks related to editing notes.  Once you have visited multiple screens and edited notes, follow the steps below to share/view the application log.
1. Open navigation drawer.
2. Tap ***Settings*** item in navigation drawer.
3. Tap ***Send logs*** item under ***More*** section.
4. Notice ***Share*** sheet is shown.
5. Choose any option in ***Share*** sheet to share/view application log.
6. Notice application log is as shown above.

### Review
Only one developer is required to review these changes, but anyone can perform the review.

### Release
These changes do not require release notes.
